Automate the uploading of the example images to dockerhub

craigyam opened the pull request:
- Generate sidecar dockerfile for example apps for testing
- Push the versioned example images to dockerhub after successful test run

craigyam committed:
Generate sidecar dockerfile for apps for examples testing.

The app Dockerfile.sidecar file has the version number hardcoded.
The examples automated testing builds the images to be pushed to
dockerhub so we want to make sure we are testing with the current
files and not what is uploaded for the previous release. Generate
the Dockerfile.sidecar in 2 ways:
1)  If testing the examples, pull the local release sidecar tar and
extract that.
2)  If not testing the examples, the build.exampleapps target will
generate the dockerfile with the current release version.

craigyam committed:
Push example images to dockerhub when creating a release.

- Update the travis yaml to run the examples tests on a release
- Push the versioned example images to dockerhub after examples
tests pass.

gicaceres reviewed (approved):
LGTM

rshriram commented on Makefile:
```diff
@@ -317,6 +322,8 @@ release.sidecar:
 	@cp openresty/*.tar.gz $(BUILDDIR)/opt/openresty_dist/
 	@tar -C $(BUILDDIR) -czf $(RELEASEDIR)/$(SIDECAR_RELEASE_NAME).tar.gz --transform 's:^./::' .
+	# Make a copy for examples testing
+	@cp $(RELEASEDIR)/$(SIDECAR_RELEASE_NAME).tar.gz $(RELEASEDIR)/a8sidecar-current.tar.gz
```
This will end up in the github public release. You can reuse the existing files in the release dir right ?

craigyam replied:
Done

rshriram commented on Makefile:
```diff
@@ -166,6 +166,11 @@ build.exampleapps:
+	@echo "--> building example apps for testing"
+	@examples/apps/helloworld/build-services.sh "$(APP_VER_ABBR)" "examples"
+	@examples/apps/bookinfo/build-services.sh "$(APP_VER_ABBR)" "examples"
+
```
What's the difference between this and the previous target?

craigyam replied:
build.exampleapps is the standalone build for the apps.  So it will generate the yamls and it will generate the images with the download links.
build.exampleappstesting is for the automated testing path.  So it will build the release tars and generate the images using the local files.

rshriram commented on examples/apps/bookinfo/build-services.sh:
```diff
@@ -27,19 +27,51 @@ if [ "$APP_VER" == "unknown" ]; then
 pushd $SCRIPTDIR/details
   docker build -t amalgam8/a8-examples-bookinfo-details-v1:$APP_VER .
-  docker build -t amalgam8/a8-examples-bookinfo-details-sidecar-v1:$APP_VER -f Dockerfile.sidecar .
+  generate_sidecar_dockerfile
```
Wouldn't it be simpler to templatize the example docker files and run the " generate dockerfile app name" command?

craigyam replied:
Modified Dockerfile.sidecar rather than generating one.

craigyam committed:
Code review comments.

- Update Dockerfile.sidecar for example apps to always use local
  files.
- Use existing sidecar tar instead of copy.
- Make the list of example images an array in push_dockerhub script.
- Allow setting version for tag in Makefile.

rshriram commented on examples/apps/bookinfo/details/Dockerfile.sidecar:
```diff
@@ -14,11 +14,20 @@
+    && tar -xzf ${A8_TMP}/opt/openresty_dist/*.tar.gz -C / \
+    && tar -xzf /opt/microservices/${A8_SIDECAR_RELEASE}.tar.gz -C / \
+    && rm -rf ${A8_TMP} \
+    && rm /opt/microservices/${A8_SIDECAR_RELEASE}.tar.gz
```
This is outdated. Please see the newer Makefile and the a8sidecar-envoy.sh

rshriram reviewed (changes requested):
Sorry for the delay. Merge conflicts and one main change to Dockerfiles (using Envoy instead of nginx).

Commit:
use envoy sidecar installation

Commit:
cleanup docker files and push major and minor tags for examples

rshriram commented:
fixes #404

rshriram commented:
@ijsnellf @GregHanson @zcahana @tomerGolany 
The DNS test is failing.. any idea why? This PR had nothing to do with the sidecar go code.

```bash
=== RUN   TestRequestsWithSubTags
time="2017-01-30T00:54:27Z" level=info msg="Starting DNS server" 
time="2017-01-30T00:54:27Z" level=error msg="Error starting DNS server" error="listen udp :8485: bind: address already in use" 
time="2017-01-30T00:54:29Z" level=info msg="Shutting down DNS server" 
time="2017-01-30T00:54:29Z" level=error msg="Error shutting down DNS server" error="dns: server not started" 
--- FAIL: TestRequestsWithSubTags (2.20s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x18 pc=0x48688c]
goroutine 15 [running]:
panic(0x777100, 0xc4200121c0)
	/home/travis/.gimme/versions/go1.7.1.linux.amd64/src/runtime/panic.go:500 +0x1a1
testing.tRunner.func1(0xc420084780)
	/home/travis/.gimme/versions/go1.7.1.linux.amd64/src/testing/testing.go:579 +0x25d
panic(0x777100, 0xc4200121c0)
	/home/travis/.gimme/versions/go1.7.1.linux.amd64/src/runtime/panic.go:458 +0x243
github.com/amalgam8/amalgam8/sidecar/dns.(*TestSuite).TestRequestsWithSubTags(0xc42001aa50)
	/home/travis/gopath/src/github.com/amalgam8/amalgam8/sidecar/dns/server_test.go:170 +0xac
reflect.Value.call(0xc42020d080, 0xc4200345d0, 0x13, 0x7d1e26, 0x4, 0xc420051f40, 0x1, 0x1, 0x5cd9f8, 0x7ceee0, ...)
	/home/travis/.gimme/versions/go1.7.1.linux.amd64/src/reflect/value.go:434 +0x5c8
reflect.Value.Call(0xc42020d080, 0xc4200345d0, 0x13, 0xc420051f40, 0x1, 0x1, 0xc420017f40, 0xbd0, 0xbe)
	/home/travis/.gimme/versions/go1.7.1.linux.amd64/src/reflect/value.go:302 +0xa4
github.com/amalgam8/amalgam8/vendor/github.com/stretchr/testify/suite.Run.func2(0xc420084780)
	/home/travis/gopath/src/github.com/amalgam8/amalgam8/vendor/github.com/stretchr/testify/suite/suite.go:95 +0x1cb
testing.tRunner(0xc420084780, 0xc42032c000)
	/home/travis/.gimme/versions/go1.7.1.linux.amd64/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/home/travis/.gimme/versions/go1.7.1.linux.amd64/src/testing/testing.go:646 +0x2ec
FAIL	github.com/amalgam8/amalgam8/sidecar/dns	2.826s
```

zcahana commented:
@rshriram this is due to a test bug in DNS unittests causing intermittent failures. It's captured in #443. For the mean time, one can just restart the Travis build which will likely pass.
